README.md
# DataHut

A small, portable data warehouse for Ruby for analytics on anything!

DataHut has basic features for small one-off analytics like parsing error logs and such.  Like its bigger cousin (the Data Warehouse) it has support for *extract*, *transform* and *load* processes (ETL).  Unlike its bigger cousin it is simple to setup and use for simple projects.

*Extract* your data from anywhere, *transform* it however you like and *analyze* it for insights!

<img src="https://raw.github.com/coldnebo/data_hut/master/samples/weather_files/screenshot.png" width="70%"/>  
*from [samples/weather_station.rb](https://github.com/coldnebo/data_hut/blob/master/samples/weather_station.rb)*


## Installation

Add this line to your application's Gemfile:

    gem 'data_hut'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install data_hut

## Usage

Setting up a datahut is easy...

    require 'data_hut'
    require 'pry'

    dh = DataHut.connect("scratch")

    data = [{name: "barney", age: 27},
            {name: "phil", age: 31},
            {name: "fred", age: 44}]

    # extract your data by iterating over your data format (from whatever source) and map it to a record model...
    dh.extract(data) do |r, d|
      r.name = d[:name]
      r.age = d[:age]
    end

    # transform your data by adding fields to it
    dh.transform do |r|
      r.eligible = r.age < 30
    end

    # operate on your dataset by using chained queries
    ds = dh.dataset

    binding.pry

DataHut provides access to the underlying [Sequel::Dataset](http://sequel.rubyforge.org/rdoc/classes/Sequel/Dataset.html) using
a Sequel::Model binding.  This allows you to query individual fields and stats from the dataset, but also returns rows as objects that are accessed with the same uniform object syntax you used for extracting and transforming... i.e.:

    [1] pry(main)> person = ds.first
    [2] pry(main)> [person.name, person.age]
    => ["barney", 27]

And here's some of the other powerful things you can do with a Sequel::Dataset:

    [2] pry(main)> ds.where(eligible: false).count
    => 2
    [3] pry(main)> ds.avg(:age)
    => 34.0
    [4] pry(main)> ds.max(:age)
    => 44
    [5] pry(main)> ds.min(:age)
    => 27

But you can also name subsets of data and work with those instead:

    [6] pry(main)> ineligible = ds.where(eligible: false)
    => #<Sequel::SQLite::Dataset: "SELECT * FROM `data_warehouse` WHERE (`eligible` = 'f')">

    [26] pry(main)> ineligible.avg(:age)
    => 37.5
    [24] pry(main)> ineligible.order(Sequel.desc(:age)).all
    => [#< @values={:dw_id=>3, :name=>"fred", :age=>44, :eligible=>false}>,
     #< @values={:dw_id=>2, :name=>"phil", :age=>31, :eligible=>false}>]

And results remain Sequel::Model objects, so you can access fields with object notation:

    [32] pry(main)> record = ineligible.order(Sequel.desc(:age)).first
    => #< @values={:dw_id=>3, :name=>"fred", :age=>44, :eligible=>false}>
    [33] pry(main)> record.name
    => "fred"
    [34] pry(main)> record.age
    => 44

Or you can output results directly to JSON. 

    [1] pry(main)> puts ds.group_and_count(:name).all.to_json
    [{"name":"barney","count":3},{"name":"fred","count":1},{"name":"phil","count":2}]

(See [samples/weather_station.rb](https://github.com/coldnebo/data_hut/blob/master/samples/weather_station.rb) for an example of using JSON output to visualize data with d3.js.)
    
Read more about the [Sequel gem](http://sequel.rubyforge.org/) to determine what operations you can perform on a DataHut dataset.

## A More Ambitious Example...

Taking a popular game like League of Legends and hand-rolling some simple analysis of the champions.  Look at the following sample
code:

* [samples/league_of_legends.rb](https://github.com/coldnebo/data_hut/blob/master/samples/league_of_legends.rb)

Running this sample scrapes some game statistics from an official website and then transforms this base data with 
extra fields containing different totals and indices that we can construct however we like. 
Now that we have some data extracted and some initial transforms defined, lets play with the results...

* who has the most base damage?

        [1] pry(main)> ds.order(Sequel.desc(:damage)).limit(5).collect{|c| {c.name => c.damage}}
        => [{"Taric"=>58.0},
         {"Maokai"=>58.0},
         {"Warwick"=>56.76},
         {"Singed"=>56.65},
         {"Poppy"=>56.3}]


* but wait a minute... what about at level 18?  Fortunately, we've transformed our data to add some extra "total" fields for each stat...

        [2] pry(main)> ds.order(Sequel.desc(:total_damage)).limit(5).collect{|c| {c.name => c.total_damage}}
        => [{"Skarner"=>129.70000000000002},
         {"Cho'Gath"=>129.70000000000002},
         {"Kassadin"=>122.5},
         {"Taric"=>121.0},
         {"Alistar"=>120.19}]

* how about using some of the indices we defined?... for instance, if we want to know which champions produce the greatest damage we could try sorting by our 'nuke_index', (notice that the assumptions on what make a good 'nuke' are subjective, but that's the fun of it; we can model our assumptions and see how the data changes in response.)

        [3] pry(main)> ds.order(Sequel.desc(:nuke_index)).limit(5).collect{|c| {c.name => [c.total_damage, c.total_move_speed, c.total_mana, c.ability_power]}}
        => [{"Karthus"=>[100.7, 335.0, 1368.0, 10]},
         {"Morgana"=>[114.58, 335.0, 1320.0, 9]},
         {"Ryze"=>[106.0, 335.0, 1240.0, 10]},
         {"Karma"=>[109.4, 335.0, 1320.0, 9]},
         {"Lux"=>[109.4, 340.0, 1150.0, 10]}]

From my experience in the game, these champions are certainly heavy hitters.  What do you think?

* and (now I risk becoming addicted to DataHut myself), here's some further guesses with an 'easy_nuke' index (champions that have a lot of damage, but are also less difficult to play):

        [4] pry(main)> ds.order(Sequel.desc(:easy_nuke_index)).limit(5).collect{|c| c.name}
        => ["Sona", "Ryze", "Nasus", "Soraka", "Heimerdinger"]

* makes sense, but is still fascinating... what about my crack at a support_index (champions that have a lot of regen, staying power, etc.)?

        [5] pry(main)> ds.order(Sequel.desc(:support_index)).limit(5).collect{|c| c.name}
        => ["Sion", "Diana", "Nunu", "Nautilus", "Amumu"]



You get the idea now!  *Extract* your data from anywhere, *transform* it however you like and *analyze* it for insights!

Have fun!


## Metadata Object Store

DataHut also supports a basic Ruby object store for storing persistent metadata that might be useful during extract and transform passes.

*Examples:*

* [samples/league_of_legends.rb](https://github.com/coldnebo/data_hut/blob/master/samples/league_of_legends.rb):

        dh.extract(urls) do |r, url|
          ...
          # names => ["damage", "health", "mana", "move_speed", "armor", "spell_block", "health_regen", "mana_regen"]

          # DataHut also allows you to store metadata for the data warehouse during any processing phase for later retrieval.
          # Since we extract the data only once, but may need stats names for subsequent transforms, we can store the 
          # stats names in the metadata for later use:
          dh.store_meta(:stats, names)
          ...
        end
        ... 
        # later... we can fetch the metadata that was written during the extract phase and use it... 
        stats = dh.fetch_meta(:stats)
        # stats => ["damage", "health", "mana", "move_speed", "armor", "spell_block", "health_regen", "mana_regen"]

**Caveats:** Because the datastore can support any Ruby object (including custom ones) it is up to the caller to make sure that custom classes are in context before storage and fetch.  i.e. if you store a custom object and then fetch it in a context that doesn't have that class loaded, you'll get an error.  For this reason it is safest to use standard Ruby types (e.g. Array, Hash, etc.) that will always be present.

See {DataHut::DataWarehouse#store_meta} and {DataHut::DataWarehouse#fetch_meta} for details.

## TODOS

* further optimizations
* time-based series and binning helpers (by week/day/hour/5-min/etc).

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# Changelog

## 0.1.1

* fix rogue 'ArgumentError: DataHut: Ruby type '...' not supported by Sequel.' errors.

## 0.1.0

* added support for jruby 1.9 mode.

## 0.0.9

* added to_json support for dataset results.

## 0.0.8 

* handle unsanitized nil values properly - If your input data has occasional nil values during extract or transform, you may have seen:
        `DataHut: Ruby type 'NilClass' not supported by Sequel...`
  DataHut now handles nil values instead of raising this exception so that it is easier to work with unsanitized datasets.

* added `DataHut::DataWarehouse#non_unique` which allows you to specify any test of uniqueness for early skipping during transform or extract phases.  DataHut has duplicate detection built-in, i.e. it doesn't allow identical records to be inserted.  However in the past, you had to wait for all the fields to be added or transformed before this detection was done.  `non-unique` allows you to define more specific uniqueness paramters for early skipping without going through all that.  i.e. you have a feed where you know a dup is some kind of GUID... simply test if the GUID is unique *before* going any further...

        dh.extract(data) do |r, d|
          next if dh.not_unique(guid: d[:guid])
          r.guid = d[:guid]
          r.name = d[:name]
          r.age = d[:age]
          ...
        end

## 0.0.7

* added capability to store and fetch arbitrary metadata from the DataHut. 

  This is useful in the case motivated by the samples/league_of_legends.rb:
    stat name is known at initial extract time, however
    subsequent transform runs may or may not have any transient variables for stat names... hence the metadata needs to be stored 
    somewhere for future transform processing.
    note: stat name is not of the same cardinality as the data records themselves, so it is truly metadata that governs how the records
    are understood.

## 0.0.6 

* externalized the Sequel database logger so that it can be set by DataHut clients.  See DataHut::DataWarehouse#logger=

* added type checking on extract and transform to ensure safe operation with underlying Sequel sqlite3 database.

## 0.0.5

* added rdoc

* added tests; 100% code coverage.

## 0.0.4

* added the capability to mark records in the datahut as processed so that transform passes can ignore previously processed data and only process new data... good for cycles where you pull regular updates and then process them.

* added capability to force the transform to write in spite of processed; good for situations where you are playing with the structure of the transform and want to regenerate the data.


## 0.0.3

* fixed an update issue found in transforms where data was written successfully, but Sequel::Model couldn't read it immediately after.


## 0.0.2

* fixed problem with multiple instances of the datahut returning only a single dataset instance.

* added more interesting example to motivate edge cases.


## 0.0.1

* initial checkin. basic functionality


lib/data_hut/data_warehouse.rb
require 'sequel'
require 'ostruct'
require 'logger'
require 'json'


module DataHut

  # The DataHut::DataWarehouse comprehensively manages all the heavy lifting of creating a data system for your analytics.
  # So during *extract* and *transform* phases you don't have to worry about the schema or the data types you'll be using... 
  # just start scraping and playing with the data extraction, DataHut will take care of introspecting your final data records
  # and creating or altering the DataHut schema for you, auto-magically.
  #
  # @example 
  #   require 'data_hut'
  #   require 'pry'   # not necessary, but very useful
  #
  #   dh = DataHut.connect("scratch")
  #   data = [{name: "barney", age: 27, login: DateTime.parse('2008-05-03') },
  #           {name: "phil", age: 31},
  #           {name: "fred", age: 44, login: DateTime.parse('2013-02-07')}]
  #
  #   # extract your data by iterating over your data format (from whatever source) and map it to a record model...
  #   dh.extract(data) do |r, d|
  #     r.name = d[:name]
  #     r.age = d[:age]
  #     # you can do anything you need to within the extract block to ensure data quality if you want:
  #     d[:login] = DateTime.new unless d.has_key?(:login)
  #     r.last_active = d[:login]
  #     print 'v'
  #   end
  #
  #   # transform your data by adding fields to it
  #   dh.transform do |r|
  #     r.eligible = r.age < 30
  #     print '*'
  #   end
  #
  #   # mark all the records as processed to avoid re-transforming them.
  #   dh.transform_complete
  #   ds = dh.dataset
  #   binding.pry   # play with ds.
  #   [1] pry(main)> ds.avg(:age)
  #   => 34.0
  #   [2] pry(main)> ineligible = ds.where(eligible: false)
  #   [3] pry(main)> ineligible.avg(:age)
  #   => 37.5 
  class DataWarehouse
    private_class_method :new

    # creates or opens an existing connection to a DataHut data store.
    #
    # @param name [String] name of the DataHut.  This will also be the name of the sqlite3 file written 
    #   to the current working directory (e.g. './<name>.db')
    # @return [DataHut::DataWarehouse] instance
    def self.connect(name)
      new(name)
    end

    # access the DataHut dataset. See {http://sequel.rubyforge.org/rdoc/classes/Sequel/Dataset.html Sequel::Dataset} 
    # for available operations on the dataset. 
    # 
    # @note the resulting [Sequel::Model] additionally supports a #to_json method for JSON export of the dataset results.
    # @return [Sequel::Model] instance bound to the data warehouse. Use this handle to query and analyze the datahut.
    def dataset
      klass = Class.new(Sequel::Model(@db[:data_warehouse]))
      klass.class_eval do
        def to_json(*a)
          values.to_json(*a)
        end
      end
      klass
    end

    # used to extract data from whatever source you wish. As long as the data forms an enumerable collection, 
    # you can pass it to extract along with a block that specifies how you which the DataHut *record* to be 
    # mapped from the source *element* of the collection.
    # 
    # @example Extracting fields from a hash and assigning it to a field on a record
    #  data = [{name: "barney", age: 27, login: DateTime.parse('2008-05-03') }]
    #  dh.extract(data) do |r, d|
    #    r.name = d[:name]
    #    r.age  = d[:age]
    #  end
    # 
    # @param data [Enumerable]
    # @yield [record, element] lets you control the mapping of data elements to record fields
    # @yieldparam record an OpenStruct that allows you to create fields dynamically on the record as needed. 
    #   These fields will automatically be added to the schema behind the DataHut using the ruby data type you assigned to the record.
    #   See {http://sequel.rubyforge.org/rdoc/files/doc/schema_modification_rdoc.html Sequel Schema Modification Methods} for 
    #   more information about supported ruby data types you can use.
    # @yieldparam element an element from your data.
    # @raise [ArgumentError] if you don't provide a block
    # @return [void]
    # @note Duplicate records (all fields and values must match) are automatically not inserted at the end of an extract iteration. You may
    #   also skip duplicate extracts early in the iteration by using {#not_unique}.
    # @note Fields with nil values in records are skipped because the underlying database defaults these to 
    #   nil already. However you must have at least one non-nil value in order for the field to be automatically created,
    #   otherwise subsequent transform layers may report errors on trying to access the field.
    def extract(data)
      raise(ArgumentError, "a block is required for extract.", caller) unless block_given?

      data.each do |d|
        r = OpenStruct.new
        yield r, d
        store(r)
      end
    end

    # used to transform data already extracted into a DataHut.  You can also use *transform* to create new synthetic data fields 
    # from existing fields.  You may create as many transform blocks (i.e. 'passes') as you like.
    #
    # @example Defining 'eligibility' based on arbitrary age criteria.
    #   dh.transform do |r|
    #     r.eligible = r.age < 30      # using extracted to create a synthetic boolean field
    #   end
    # 
    # @param forced if set to 'true', this transform will iterate over records already marked processed.  This can be useful for 
    #   layers of transforms that deal with analytics where the analytical model may need to rapidly change as you explore the data.
    #   See the second transform in {https://github.com/coldnebo/data_hut/blob/master/samples/league_of_legends.rb#L102 samples/league_of_legends.rb:102}.
    # @yield [record] lets you modify the DataHut record
    # @yieldparam record an OpenStruct that fronts the DataHut record.  You may access existing fields on this record or create new 
    #   fields to store synthetic data from a transform pass. 
    #   These fields will automatically be added to the schema behind the DataHut using the ruby data type you assigned to the record.
    #   See {http://sequel.rubyforge.org/rdoc/files/doc/schema_modification_rdoc.html Sequel Schema Modification Methods} for 
    #   more information about supported ruby data types you can use.
    # @raise [ArgumentError] if you don't provide a block
    # @return [void]
    def transform(forced=false)
      raise(ArgumentError, "a block is required for transform.", caller) unless block_given?

      # now process all the records with the updated schema...
      @db[:data_warehouse].each do |h|
        # check for processed if not forced
        unless forced
          next if h[:dw_processed] == TRUE_VALUE
        end
        # then get rid of the internal id and processed flags
        dw_id = h.delete(:dw_id)
        h.delete(:dw_processed)
        # copy record fields to an openstruct
        r = OpenStruct.new(h)
        # and let the transformer modify it...
        yield r
        # get the update hash from the openstruct
        h = ostruct_to_hash(r)
        # now add any new transformation fields to the schema...
        adapt_schema(h)
        # and use it to update the record
        @db[:data_warehouse].where(dw_id: dw_id).update(h)
      end
    end

    # marks all the records in the DataHut as 'processed'.  Useful as the last command in a sequence of extract and transform passes.
    # 
    # @example a simple log analysis system (pseudocode)
    #   rake update
    #      extract apache logs  (only adds new logs since last update)
    #      transform logs into types of response (error, ok, met_SLA (service level agreement, etc.))  (only transforms unprocessed (new) logs)
    #      transform_complete (marks the update complete)
    #      dh.dataset is used to visualize graphs with d3.js
    #   end
    #
    # @return [void]
    def transform_complete
      @db[:data_warehouse].update(:dw_processed => true)
    end

    # attach a Logger to the underlying Sequel database so that you can debug or monitor database actions.
    # See {http://sequel.rubyforge.org/rdoc/classes/Sequel/Database.html#method-i-logger-3D Sequel::Database#logger=}.
    #
    # @example 
    #   dh.logger = Logger.new(STDOUT)
    #
    # @param logger [Logger] a logger for the underlying Sequel actions.
    # @raise [ArgumentError] if passed a logger that is not a kind of {http://www.ruby-doc.org/stdlib-1.9.3//libdoc/logger/rdoc/Logger.html Logger}.
    # @return [void]
    def logger=(logger)
      raise(ArgumentError, "logger must be a type of Logger.") unless logger.kind_of?(Logger)
      @db.logger = logger
    end

    # stores any Ruby object as metadata in the datahut.
    #
    # @param key [Symbol] to reference the metadata by
    # @param value [Object] ruby object to store in metadata
    # @return [void]
    # @note Because the datastore can support any Ruby object (including custom ones) it is up to 
    #   the caller to make sure that custom classes are in context before storage and fetch.  i.e. if you 
    #   store a custom object and then fetch it in a context that doesn't have that class loaded, you'll get an error.  
    #   For this reason it is safest to use standard Ruby types (e.g. Array, Hash, etc.) that will always be present.
    def store_meta(key, value)
      key = key.to_s if key.instance_of?(Symbol)
      begin 
        value = Sequel::SQL::Blob.new(Marshal.dump(value))
        if (@db[:data_warehouse_meta].where(key: key).count > 0)
          @db[:data_warehouse_meta].where(key: key).update(value: value)
        else
          @db[:data_warehouse_meta].insert(key: key, value: value)
        end
      rescue Exception => e
        raise(ArgumentError, "DataHut: unable to store metadata value #{value.inspect}: #{e.message}", caller)
      end
    end

    # retrieves any Ruby object stored as metadata.
    #
    # @param key [Symbol] to lookup the metadata by
    # @return [Object] ruby object that was fetched from metadata
    # @note Because the datastore can support any Ruby object (including custom ones) it is up to 
    #   the caller to make sure that custom classes are in context before storage and fetch.  i.e. if you 
    #   store a custom object and then fetch it in a context that doesn't have that class loaded, you'll get an error.  
    #   For this reason it is safest to use standard Ruby types (e.g. Array, Hash, etc.) that will always be present.
    def fetch_meta(key)
      key = key.to_s if key.instance_of?(Symbol)
      begin
        r = @db[:data_warehouse_meta].where(key: key).first
        value = r[:value] unless r.nil?
        value = Marshal.load(value) unless value.nil?
      rescue Exception => e
        raise(RuntimeError, "DataHut: unable to fetch metadata key #{key}: #{e.message}", caller)
      end
      value
    end

    # used to determine if the specified fields and values are unique in the datahut.
    # 
    # @example
    #   dh.extract(data) do |r, d|
    #     next if dh.not_unique(name: d[:name])
    #     r.name = d[:name]
    #     r.age = d[:age]
    #     ...
    #   end
    #
    # @note exactly duplicate records are automatically skipped at the end of an extract iteration (see {#extract}). This 
    #   method is useful if an extract iteration takes a long time and you want to skip duplicates early in the iteration.
    # @param hash [Hash] of the key, value pairs specifying a partial record by which to consider records unique.
    # @return [Boolean] true if the {field: value} already exists, false otherwise (including if the column doesn't yet exist.)
    def not_unique(hash)
      @db[:data_warehouse].where(hash).count > 0 rescue false
    end

    private 

    # for some reason, jdbc-sqlite3 unmarshals TrueClass columns as Fixnums, so make this a plaform dependency.
    TRUE_VALUE = (RUBY_PLATFORM == "java") ? 1 : true
  
    def initialize(name)
      @db_file = "#{name}.db"
      
      if RUBY_PLATFORM == "java"
        @db = Sequel.connect("jdbc:sqlite:#{@db_file}")
      else
        @db = Sequel.sqlite(@db_file)
      end 
      
      unless @db.table_exists?(:data_warehouse)
        @db.create_table(:data_warehouse) do
          primary_key :dw_id
          column :dw_processed, TrueClass, :null => false, :default => false
        end
      end

      unless @db.table_exists?(:data_warehouse_meta)
        @db.create_table(:data_warehouse_meta) do
          primary_key :dw_id
          String :key
          index :key
          blob :value
        end
      end
    end

    def store(r)
      h = ostruct_to_hash(r)
      adapt_schema(h)
      # don't insert dups
      unless not_unique(h)
        @db[:data_warehouse].insert(h)
      end
    end

    def ostruct_to_hash(r)
      h = r.marshal_dump
      h.reject{|k,v| v.nil?}  # you can't define a column type "NilClass", so strip these before adapting the schema
    end

    def adapt_schema(h)
      h.keys.each do |key|
        type = h[key].class
        unless Sequel::Schema::CreateTableGenerator::GENERIC_TYPES.include?(type.to_s)
          raise(ArgumentError, "DataHut: Ruby type '#{type}' not supported by Sequel. Must be one of the supported types: #{Sequel::Schema::CreateTableGenerator::GENERIC_TYPES.inspect}", caller)
        end
        unless @db[:data_warehouse].columns.include?(key)
          @db.alter_table(:data_warehouse) do 
            add_column key, type
            add_index key
          end
        end
      end
    end

  end
end